build-tools/packages/build-cli/src/commands/generate/releaseNotes.ts

```diff
@@ -0,0 +1,202 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "node:assert";
+import { writeFile } from "node:fs/promises";
+import path from "node:path";
+import { loadFluidBuildConfig } from "@fluidframework/build-tools";
+import { Flags } from "@oclif/core";
+import { StringBuilder } from "@rushstack/node-core-library";
+import { format as prettier } from "prettier";
+import { remark } from "remark";
+import remarkGfm from "remark-gfm";
+import remarkGithub, { defaultBuildUrl } from "remark-github";
+import admonitions from "remark-github-beta-blockquote-admonitions";
+import remarkToc from "remark-toc";
+
+import { releaseGroupFlag } from "../../flags.js";
+import {
+	BaseCommand,
+	DEFAULT_CHANGESET_PATH,
+	FluidCustomChangeSetMetadataDefaults,
+	UNKNOWN_SECTION,
+	difference,
+	groupBySection,
+	loadChangesets,
+} from "../../library/index.js";
+
+/**
+ * Generates release notes from individual changeset files.
+ */
+export default class GenerateReleaseNotesCommand extends BaseCommand<
+	typeof GenerateReleaseNotesCommand
+> {
+	static readonly summary = `Generates release notes from individual changeset files.`;
+
+	// Enables the global JSON flag in oclif.
+	static readonly enableJsonFlag = true;
+
+	static readonly flags = {
+		releaseGroup: releaseGroupFlag({
+			required: true,
+		}),
+		releaseType: Flags.custom<"major" | "minor">({
+			char: "t",
+			description: "The type of release for which the release notes are being generated.",
+			options: ["major", "minor"],
+			required: true,
+			parse: async (input) => {
+				if (input === "major" || input === "minor") {
+					return input;
+				}
+
+				throw new Error(`Invalid release type: ${input}`);
+			},
+		})(),
+		out: Flags.file({
+			description: `Output the results to this file.`,
+			required: true,
+			default: "RELEASE_NOTES.md",
+			// exists: true
+		}),
+		...BaseCommand.flags,
+	} as const;
+
+	static readonly examples = [
+		{
+			description: `Generate release notes for a minor release of the client release group.`,
+			command: "<%= config.bin %> <%= command.id %> -g client -t minor",
+		},
+		{
+			description: `You can output a different file using the --out flag.`,
+			command:
+				"<%= config.bin %> <%= command.id %> -g client -t minor --out RELEASE_NOTES/2.1.0.md",
+		},
+	];
+
+	public async run(): Promise<string> {
+		const context = await this.getContext();
+		const { flags, logger } = this;
+
+		const releaseGroup = context.repo.releaseGroups.get(flags.releaseGroup);
+		if (releaseGroup === undefined) {
+			this.errorLog(`Unknown release group: ${flags.releaseGroup}`);
+			this.exit(2);
```

Comment:
Nit, is there an advantage to this over just ``throw new Error(`Unknown release group: ${flags.releaseGroup}\`)`` and let the process exit due to an unhandled error? Does oclif handle all errors in a non-intuitive way at the top level? Also in line 111 we use `assert()` instead of a check like this, also with the intention of crashing the process, I believe. If that works fine, I'd say we standardize on `assert()` for process-terminating checks.

Tangent: just learned you can make backticks appear inside inline code blocks by using double-backticks instead of single ones around the block.

Reply:
This goes through the "managed exit" path vs. the "unhandled exception/crash" path. In general if we know we're exiting
because of a known condition, it's clearer to explicitly exit. It's not an "exception"al case - it's a known invalid
case.

> Also in line 111 we use assert() instead of a check like this

That assert isn't needed. Oclif will ensure that flag is provided and has a value. There are times, however, when
TypeScript "forgets" that a flag value isn't nullable. That's when we use the assert pattern, which results in less
emitted code than other options (except `!` but the assert is clearer IMO).

Reply:
To be clear I removed the unneeded assert in this case.

Reply:
Not really arguing against it (even though my gut feeling would be for centralizing the exit logic), but I find it interesting that @ChumpChief just went the other way with a change in stress tests, consolidating error logging and exits at the top of the main file. Tagged in case he has comments. Not blocking for me.

Reply:
I don't have extensive thoughts in this space, but I saw some parallels to the typical discussions about the negatives of `throw` - that it provides control flow in the program that is hard to follow or expect (an "invisible exit" to the caller) - with this mentality, `exit` is yet another variation of the problem.  You could also think of it as being a `GOTO end` equivalent, and therefore maybe considered harmful depending on your perspective :)

In the stress tests, this specifically resulted in `exit` before appropriate logging could happen, for example.

So my hot take with not too much thought is that it seems nice to:
1. Avoid both `throw` and `exit` whenever possible ("don't use exceptions for control flow")
2. When throwing, allow the throw to propagate up to whichever caller makes sense to be responsible for handling it (e.g. don't swallow the error locally like the stress tests were doing - tell the caller you screwed up if they ought to know)
3. Exit at the top in all cases.  Error exits happen adjacent to non-error exits.  If the program exits in a single place then it is easier to inspect and audit exit conditions and how they happen.  At the top, it's reasonable to expect you also have a bird's eye view of any parallelized unresolved async stuff - this can be permitted to finish up or be aborted before the process exits, etc.

Reply:
> Avoid both throw and exit whenever possible ("don't use exceptions for control flow")

I don't think it is feasible to not call exit in this context, either explicitly or [implicitly via this.error](https://oclif.io/docs/commands#thiserrormessage-string--error-options-code-string-exit-number-ref-string-suggestions-string). It's the standard oclif pattern.

That said, this is _not_ process.exit(). That eventually gets called, but from a central location in the oclif infrastructure. More at https://oclif.io/docs/error_handling

Reply:
No worries, when in Rome :) Take my comments as just general findings I had while working on the stress tests that may be more or less applicable depending on what you're doing.